RPP: Legacy Container

RadoslavGeorgiev opened the pull request:
Partially fixes #6707, specifically this part:

> Figure out a way to bring `includes` classes into the container. Hard-coding methods [like this](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Proxies/LegacyProxy.php#L69-L71) seems suitable.

### Changes proposed in this Pull Request

Building on top of https://github.com/Automattic/woocommerce-payments/pull/6862, this PR tries to allow classes from the legacy `includes` directory to be loaded as dependencies by `src` classes. This is achieved through `WCPay\Internal\DependencyManagement\DelegateContainer\LegacyContainer`.

`LegacyContainer` implements the PSR-11 interface, and is then used as a delegate container within `WCPay\Container`. It works as a hard-coded bridge between `src` and `includes`.

Essentially, it:

1. Transforms the name of the needed class to a method. For example `WCPay\Mode` becomes `get_wcpay_core_mode_instance`.
2. Checks if the class has such a method, and uses this check within the public `has()` method, which indicates whether the container can provide a certain class.
3. Uses the method to provide the instance through `get()`.

Right now `WCPay\Mode` is the only example here, but it can be used for all sorts of classes, ex. `WC_Payments_API_Client` from [this question](https://github.com/Automattic/woocommerce-payments/pull/6862#discussion_r1281826677):

```php
private function get_wc_payments_api_client_instance() {
	return \WC_Payments::get_payments_api_client();
}
```

#### Potential classes

Assuming this concept gets approved, before merging this PR, the container should include all classes, available through `WC_Payments::get_[CLASS_NAME]`.

#### Alternatives

What I've considered:

1. Using Woo's `LegacyProxy` ([ref](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Proxies/LegacyProxy.php)): We could use `LegacyProxy` as a dependency, and then `$legacy_proxy->get_instance_of()`, together with hard-coded methods for specific classes ([example](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Proxies/LegacyProxy.php#L69)), similarly to the `LegacyContainer` here. The upside of this approach is that `src` code does not depend on `includes` code, the downside is the extra layer. I don't see a reason for introducing it as an obstacle.
2. The legacy `WC_Payments` class could feed dependencies into the container, but this would mean allowing external modifications of the League container. That's also something that should be avoided IMO.

### Testing instructions

There is not much to manually test, as we've got few `src` classes and instances where they have dependencies.

You can add this snippet somewhere in your code, and open the front-end:

```php
add_action( 'template_redirect', function() {
	$service = wcpay_get_container()->get( ExampleServiceWithDependencies::class );
	var_dump( $service );
	exit;
} );
```

If instantiation succeeds, this means that `LegacyContainer` is working, and the WooPayments container is delegating to it.

I also recommend looking into the additional container tests.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests

htdat committed:
Add PaymentService and index files

htdat committed:
Exclude src from WP filename

htdat committed:
Add folder `src` to npm run release

htdat committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Adding the autoloading configuration

htdat committed:
Add changelog

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Removing test left-overs

htdat committed:
Change to PaymentProcessingService, use singular for Service

RadoslavGeorgiev committed:
First changes

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Update changelog/rpp-6677-psr-autoloader

Co-authored-by: Dat Hoang <htdat@users.noreply.github.com>

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge fix

RadoslavGeorgiev committed:
Switching to a unit test instead of a manual one

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Introducing the `lib` directory.

Because Mozart is a dev dependency, and not available within actions. This means that the lib directory will only be manipulated locally.

RadoslavGeorgiev committed:
Giving lib its own Composer instance

RadoslavGeorgiev committed:
Making psalm work.

The way we load files for Psalm prevents woocommerce-payments.php from defining functions, which would be otherwise available.

RadoslavGeorgiev committed:
Adding service providers and tests

RadoslavGeorgiev committed:
Tests part 1: Adding `wcpay_get_test_container()`.

RadoslavGeorgiev committed:
Allowing replacements during tests.

RadoslavGeorgiev committed:
Adding the container exception.

RadoslavGeorgiev committed:
Minor cleanups.

RadoslavGeorgiev committed:
Updating the lib readme

RadoslavGeorgiev committed:
Initial addition of the LegacyContainer, HooksProxy, and LegacyProxy.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6876` or branch name `add/6707-container-additions` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/6707-container-additions) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 11aa13bbad4fc461b912967c6e16d8acaa66ad24
- Build time: 2023-08-08 13:39:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev committed:
Fixing some tests.

RadoslavGeorgiev committed:
Switching to an example service for tests.

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
WIP: Replacements for delegates.

RadoslavGeorgiev committed:
Simplifying replacements

Instead of manipulating the complicated inner Container, ExtendedContainer now stores replacements locally.

RadoslavGeorgiev committed:
Fixing a unit test mistake

RadoslavGeorgiev committed:
Fixing Psalm issues

RadoslavGeorgiev committed:
Moving the Woo delegate container to a separate directory

RadoslavGeorgiev committed:
Testin improvement, incl. replacements.

Instead of overwriting existing definitions, ExtendedContainer adds a layer of replacements on top of the League container. This way references are kept after reset.

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Fixing lints

RadoslavGeorgiev committed:
Removing proxies, focusing the PR on the legacy container.

RadoslavGeorgiev committed:
Better tests:

- Adding a new `ExampleServiceWithDependencies` class, which tests all types of dependencies. `ExampleService` is just a placeholder now.
- Switched delegates with mocks for the containe tests. This way it's unit tests, rather than integration ones.
- Added a new test, which simulates all sorts of dependencies (`src, `includes`, and WooCommerce).

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' of github.com:Automattic/woocommerce-payments into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Switching namespaces

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Replacing WooPayments with WCPay

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Docs

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Switching to 4 spaces within composer.json

RadoslavGeorgiev committed:
Fixing an example

RadoslavGeorgiev committed:
Moving the initialization of the container within wcpay_get_container

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Use the container for the payment processing service

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Adding tests for the legacy container, and making it final.

RadoslavGeorgiev committed:
Adding docs

RadoslavGeorgiev committed:
Fixing all the broken tests

RadoslavGeorgiev committed:
Adding the available get_* instances from WC_Payments

RadoslavGeorgiev commented on src/Internal/Service/ExampleService.php:
```diff
@@ -0,0 +1,19 @@
+ * DI-related functionality until there are better services
+ * to test with.
+ */
+class ExampleService {
```
Yes, I have it in mind. As soon as there are a couple of real-world classes to use, they will replace the example service. I'm confident that it will happen within the payment process re-engineering project.

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-container-additions

RadoslavGeorgiev committed:
Changelog

htdat commented on src/Internal/DependencyManagement/DelegateContainer/LegacyContainer.php:
```diff
@@ -0,0 +1,151 @@
+/**
+ * WooPayments Legacy Container Delegate.
+ *
+ * This class is a proxy between `src` and `includes`, allwing
```
```suggestion
 * This class is a proxy between `src` and `includes`, allowing
```

RadoslavGeorgiev replied:
Fixed in in https://github.com/Automattic/woocommerce-payments/pull/6876/commits/31fb9dff15f5d320564c7b5914b0b7d4cb87e135

htdat commented on src/Internal/DependencyManagement/DelegateContainer/LegacyContainer.php:
```diff
@@ -0,0 +1,151 @@
+	 * Finds an entry of the container by its identifier and returns it.
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return mixed Entry.
```
May use dynamic return type here too? 
See more https://github.com/Automattic/woocommerce-payments/pull/6862#discussion_r1284425974

RadoslavGeorgiev replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/6876/commits/31fb9dff15f5d320564c7b5914b0b7d4cb87e135

htdat commented on changelog/add-6707-container-additions:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: dev
+
+Adding a LegacyContainer to `src` to allow loading classes from `includes`.
```
```suggestion
Add LegacyContainer to `src` to allow loading classes from `includes`.
```

RadoslavGeorgiev replied:
Fixed in in https://github.com/Automattic/woocommerce-payments/pull/6876/commits/31fb9dff15f5d320564c7b5914b0b7d4cb87e135

htdat commented on src/Internal/DependencyManagement/DelegateContainer/LegacyContainer.php:
```diff
@@ -0,0 +1,151 @@
+	 * @return mixed Entry.
+	 */
+	public function get( $id ) {
+		$method = $this->transform_class_to_method( $id );
+		return $this->$method();
```
I don't think we would blindly call `$this-> method()` here. 

Instead, use `$this->has($id)` to check before calling it, and then throw an exception `WCPay\Internal\DependencyManagement\ContainerException` in case `has` method return `false`.

RadoslavGeorgiev replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/6876/commits/11aa13bbad4fc461b912967c6e16d8acaa66ad24. FWIW, this container would only be called by the main container, and that one already knows not to use `get` before checking `has`, but additional safety could never hurt.

htdat reviewed (commented):
Left some comments. Other than that, everything looks great.

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-container-additions

RadoslavGeorgiev committed:
Fixing typos and adding the correct return type.

RadoslavGeorgiev committed:
Throwing exceptions within get()
